lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -214,6 +214,43 @@ TypeChecker::lookupUnqualifiedType(DeclContext *dc, DeclName name,
                                    NameLookupOptions options) {
   SmallVector<TypeDecl *, 1> decls;
 
+  // CodingKeys is a special type. Although not a protocol requirement of the
+  // Encodable/Decodable protocols, it is synthesized as part of
+  // Encodable/Decodable conformance. Unqualified lookup may occur before the
+  // type is synthesized, though, in which case, we need to make sure it gets
+  // synthesized.
+  //
+  // If this is what we're looking for, then we need to attempt to cause the
+  // type to be synthesized before performing the lookup.
+  if (name.isSimpleName(Context.Id_CodingKeys)) {
+    // UnqualifiedLookup below has no knowledge of type checking and has no way
+    // to expand protocol conformance, which is what we'd normally need.
+    //
+    // Instead, we need to iterate up the AST first here, attempting to perform
+    // qualified lookups first.
+    auto *currentDC = dc;
+    while (currentDC) {
+      auto *nominal =
+        currentDC->getAsNominalTypeOrNominalTypeExtensionContext();
+      if (nominal) {
+        // We've found a nominal context to look in. Let's preform a qualified
+        // lookup here (which will cause CodingKeys synthesis if possible).
+        auto results = lookupMemberType(nominal->getDeclContext(),
+                                        nominal->getDeclaredInterfaceType(),
+                                        Context.Id_CodingKeys, options);
+        if (results.size() > 0) {
+          for (auto &result : results) {
+            decls.push_back(result.first);
+          }
+
+          return decls;
```

Comment:
…this might not be correct—there might be a *local* type or value named CodingKeys (for whatever reason), in a function body contained within a type that conforms to Codable, and then that one should win.

I think the best answer is probably to still make this a form of recovery, but I'm having trouble coming up with the right condition because of terrible examples like this:

```swift
struct Outer {
  typealias CodingKeys = Int
  struct Inner: Codable {
    static func test() {
      print(CodingKeys.self) // should be Inner.CodingKeys
    }
  }
}

typealias CodingKeys = Float
struct Another: Codable {
  static func test() {
    print(CodingKeys.self) // should be Another.CodingKeys
  }
}
```